File: context_pipe/shadow.py
# SPDX-License-Identifier: Apache-2.0
# Copyright (c) 2026 Luis Kobayashi. All rights reserved.
"""
context_pipe/shadow.py  Shadow tool discovery.
Lists all available context-processing capabilities: configured pipes from
pipes.json and well-known CLI tools found on PATH.
"""
import os
import json
import logging
import shutil
from typing import Any

logger = logging.getLogger(__name__)

# Curated CLI tools to probe on PATH.
_KNOWN_PATH_TOOLS: dict[str, str] = {
    "jq": "Command-line JSON processor",
    "yq": "Command-line YAML/JSON/XML processor",
    "markitdown": "Converts Office/PDF/HTML documents to Markdown",
    "pandoc": "Universal document format converter",
    "rg": "Fast line-oriented search (ripgrep)",
    "fd": "Fast file finder (fd-find)",
    "bat": "Syntax-highlighted cat replacement",
}


def list_shadow_tools(config_path: str = "pipes.json") -> list[dict[str, Any]]:
    """
    Returns a combined list of available context-processing tools.
    Entries from ``pipes.json`` come first (source ``"pipes.json"``), followed
    by discovered CLI tools on PATH (source ``"PATH"``).
    Never raises  all errors are logged to stderr and an empty/partial list
    is returned.
    Args:
        config_path: Path to the project-level pipes.json file.
    Returns:
        List of tool descriptor dicts with keys:
        ``name``, ``source``, ``description``, ``nodes``.
    """
    tools: list[dict[str, Any]] = []

    # --- 1. Configured pipes from pipes.json ---
    try:
        with open(config_path, "r", encoding="utf-8") as fh:
            config = json.load(fh)
        for pipe in config.get("pipes", []):
            nodes_desc = []
            for n in pipe.get("nodes", []):
                if n.get("cmd"):
                    nodes_desc.append(n["cmd"])
                elif n.get("server") and n.get("tool"):
                    nodes_desc.append(f"mcp:{n['server']}/{n['tool']}")
                else:
                    nodes_desc.append(str(n))
            tools.append(
                {
                    "name": pipe.get("name", ""),
                    "source": "pipes.json",
                    "description": pipe.get("description", ""),
                    "nodes": nodes_desc,
                }
            )
    except FileNotFoundError:
        logger.debug("pipes.json not found at '%s'; skipping configured pipes.", config_path)
    except Exception as exc:  # noqa: BLE001
        logger.warning("Could not read pipes.json: %s", exc)

    # --- 2. Registered MCP Servers from pipes.json ---
    try:
        if os.path.exists(config_path):
            with open(config_path, "r", encoding="utf-8") as fh:
                config = json.load(fh)
            servers = config.get("servers", {})
            for name, srv_config in servers.items():
                if name.startswith("_"):  # Skip comments/examples
                    continue
                tools.append(
                    {
                        "name": name,
                        "source": "pipes.json",
                        "description": srv_config.get("description", "Registered MCP server. Can be run in custom pipes."),
                        "nodes": [],
                    }
                )
    except Exception as exc:
        logger.debug("Could not read MCP servers from config: %s", exc)

    # --- 3. Curated CLI tools on PATH ---
    for cmd, description in _KNOWN_PATH_TOOLS.items():
        if shutil.which(cmd) is not None:
            tools.append(
                {
                    "name": cmd,
                    "source": "PATH",
                    "description": description,
                    "nodes": [cmd],
                }
            )
    return tools

File: crates/cpipe/src/shadow.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 Luis Kobayashi. All rights reserved.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use crate::config::load_config_file;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct DiscoveredTool {
    pub name: String,
    pub source: String,
    pub description: String,
    pub nodes: Vec<String>,
}

lazy_static::lazy_static! {
    static ref KNOWN_PATH_TOOLS: HashMap<&'static str, &'static str> = {
        let mut m = HashMap::new();
        m.insert("jq", "Command-line JSON processor");
        m.insert("yq", "Command-line YAML/JSON/XML processor");
        m.insert("markitdown", "Converts Office/PDF/HTML documents to Markdown");
        m.insert("pandoc", "Universal document format converter");
        m.insert("rg", "Fast line-oriented search (ripgrep)");
        m.insert("fd", "Fast file finder (fd-find)");
        m.insert("bat", "Syntax-highlighted cat replacement");
        m
    };
}

pub fn which(cmd: &str, path_env: Option<&str>) -> Option<PathBuf> {
    let path_val = path_env
        .map(|s| s.to_string())
        .or_else(|| std::env::var("PATH").ok())?;
    
    let separator = if cfg!(windows) { ";" } else { ":" };
    let paths = path_val.split(separator);
    
    for p in paths {
        let dir = PathBuf::from(p);
        let exe_candidates = if cfg!(windows) {
            vec![
                dir.join(cmd),
                dir.join(format!("{}.exe", cmd)),
                dir.join(format!("{}.bat", cmd)),
                dir.join(format!("{}.cmd", cmd))
            ]
        } else {
            vec![dir.join(cmd)]
        };
        
        for candidate in exe_candidates {
            if candidate.is_file() {
                #[cfg(unix)]
                {
                    use std::os::unix::fs::PermissionsExt;
                    if let Ok(metadata) = candidate.metadata() {
                        if metadata.permissions().mode() & 0o111 != 0 {
                            return Some(candidate);
                        }
                    }
                }
                #[cfg(not(unix))]
                {
                    return Some(candidate);
                }
            }
        }
    }
    None
}

pub fn list_shadow_tools(config_path: Option<&Path>) -> Vec<DiscoveredTool> {
    let mut tools = Vec::new();
    
    // 1. Configured pipes and servers
    if let Some(path) = config_path {
        if path.exists() {
            if let Ok(config) = load_config_file(path) {
                // Configured pipes
                for pipe in config.pipes {
                    let mut nodes_desc = Vec::new();
                    for n in &pipe.nodes {
                        if n.node_type == "mcp" {
                            let server = n.server.as_deref().unwrap_or("unknown");
                            let tool = n.tool.as_deref().unwrap_or("unknown");
                            nodes_desc.push(format!("mcp:{}/{}", server, tool));
                        } else if !n.cmd.is_empty() {
                            nodes_desc.push(n.cmd.clone());
                        } else {
                            nodes_desc.push(format!("{:?}", n));
                        }
                    }
                    tools.push(DiscoveredTool {
                        name: pipe.name,
                        source: "pipes.json".to_string(),
                        description: pipe.description,
                        nodes: nodes_desc,
                    });
                }

                // Configured servers
                let mut server_names: Vec<&String> = config.servers.keys().collect();
                server_names.sort();
                for name in server_names {
                    if name.starts_with('_') {
                        continue;
                    }
                    let srv_cfg = &config.servers[name];
                    let desc = srv_cfg.description.as_deref().unwrap_or("Registered MCP server. Can be run in custom pipes.");
                    tools.push(DiscoveredTool {
                        name: name.clone(),
                        source: "pipes.json".to_string(),
                        description: desc.to_string(),
                        nodes: Vec::new(),
                    });
                }
            }
        }
    }
    
    // 2. Curated CLI tools on PATH
    let path_val = std::env::var("PATH").ok();
    let path_ref = path_val.as_deref();
    
    // Sort keys to maintain deterministic output order
    let mut keys: Vec<&&str> = KNOWN_PATH_TOOLS.keys().collect();
    keys.sort();
    
    for &cmd in keys {
        if which(cmd, path_ref).is_some() {
            if let Some(&desc) = KNOWN_PATH_TOOLS.get(cmd) {
                tools.push(DiscoveredTool {
                    name: cmd.to_string(),
                    source: "PATH".to_string(),
                    description: desc.to_string(),
                    nodes: vec![cmd.to_string()],
                });
            }
        }
    }
    
    tools
}
